# Cleaning the SonarQube Issue Data
The goal of this script is to extract the information necessary to use the SonarQube issue tags as a label for prediction.

In [1]:
import pandas as pd
import numpy as np
import os
import re

In [2]:
# data import
current_dir = os.getcwd()

# construct path to the project data folder
data_dir = os.path.join(current_dir, '..', '..', 'Data','Sonar_Issues')

# load SonarQube measure data
df1 = pd.read_csv(os.path.join(data_dir, 'SONAR_ISSUES_v1.csv'))
df2 = pd.read_csv(os.path.join(data_dir, 'SONAR_ISSUES_v2.csv'))

In [3]:
df1

projectID          creationDate             closeDate  \
0        commons-daemon  2003-09-04T23:28:19Z                   NaN   
1        commons-daemon  2003-09-04T23:28:19Z  2010-03-15T08:09:26Z   
2        commons-daemon  2003-09-04T23:28:19Z  2010-03-15T08:09:26Z   
3        commons-daemon  2003-09-04T23:28:19Z  2010-03-15T08:09:26Z   
4        commons-daemon  2003-09-04T23:28:19Z  2010-03-15T08:09:26Z   
...                 ...                   ...                   ...   
1941503     commons-cli  2017-07-29T14:36:30Z                   NaN   
1941504     commons-cli  2017-07-29T14:36:30Z                   NaN   
1941505     commons-cli  2017-07-29T14:36:30Z                   NaN   
1941506     commons-cli  2018-02-26T17:23:40Z                   NaN   
1941507     commons-cli  2018-02-26T17:23:40Z                   NaN   

                               creationCommitHash  \
0        d3416d3a25b16da3d18b3849522fa96183918e5b   
1        d3416d3a25b16da3d18b3849522fa96183918e5b   
2        d3416d3a25b16da3d18b3849522fa96183918e5b   
3        d3416d3a25b16da3d18b3849522fa96183918e5b   
4        d3416d3a25b16da3d18b3849522fa96183918e5b   
...                                           ...   
1941503  3bc9b84d3ae252800eb234a7b41981a9fef8696d   
1941504  3bc9b84d3ae252800eb234a7b41981a9fef8696d   
1941505  3bc9b84d3ae252800eb234a7b41981a9fef8696d   
1941506  b0024d482050a08efc36c3cabee37c0af0e57a10   
1941507  b0024d482050a08efc36c3cabee37c0af0e57a10   

                                  closeCommitHash        type         squid  \
0                                             NaN  CODE_SMELL  squid:S00112   
1        6cbc872eb202dfc27f2eb59b02d953c3deca32c8  CODE_SMELL  squid:S00122   
2        6cbc872eb202dfc27f2eb59b02d953c3deca32c8  CODE_SMELL  squid:S00122   
3        6cbc872eb202dfc27f2eb59b02d953c3deca32c8  CODE_SMELL  squid:S00122   
4        6cbc872eb202dfc27f2eb59b02d953c3deca32c8  CODE_SMELL  squid:S00122   
...                                           ...         ...           ...   
1941503                                       NaN  CODE_SMELL   squid:S1192   
1941504                                       NaN  CODE_SMELL   squid:S1192   
1941505                                       NaN  CODE_SMELL   squid:S1192   
1941506                                       NaN  CODE_SMELL  squid:S00112   
1941507                                       NaN  CODE_SMELL  squid:S00100   

                                                 component severity  \
0        org.apache:daemon:src/main/java/org/apache/com...    MAJOR   
1        org.apache:deamon:src/java/org/apache/commons/...    MINOR   
2        org.apache:deamon:src/java/org/apache/commons/...    MINOR   
3        org.apache:deamon:src/java/org/apache/commons/...    MINOR   
4        org.apache:deamon:src/java/org/apache/commons/...    MINOR   
...                                                    ...      ...   
1941503  org.apache:commons-cli:src/test/java/org/apach...    MINOR   
1941504  org.apache:commons-cli:src/test/java/org/apach...    MINOR   
1941505  org.apache:commons-cli:src/test/java/org/apach...    MINOR   
1941506  org.apache:commons-cli:src/test/java/org/apach...    MAJOR   
1941507  org.apache:commons-cli:src/test/java/org/apach...    MINOR   

                        project  startLine  endLine resolution  status  \
0             org.apache:daemon       71.0     71.0        NaN    OPEN   
1             org.apache:daemon      265.0    265.0      FIXED  CLOSED   
2             org.apache:daemon      259.0    259.0      FIXED  CLOSED   
3             org.apache:daemon      261.0    261.0      FIXED  CLOSED   
4             org.apache:daemon      278.0    278.0      FIXED  CLOSED   
...                         ...        ...      ...        ...     ...   
1941503  org.apache:commons-cli       34.0     34.0        NaN    OPEN   
1941504  org.apache:commons-cli       36.0     36.0        NaN    OPEN   
1941505  org.apache:commons-cli       35.0     35.0     

In [4]:
df2

PROJECT_ID CREATION_ANALYSIS_KEY             ISSUE_KEY  \
0          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWeF1RcPC4KKKThcBEZ8   
1          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWeF1RcRC4KKKThcBEZ9   
2          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWeF1RcSC4KKKThcBEZ-   
3          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWeF1RcSC4KKKThcBEZ_   
4          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWd5_rilC4KKKThc-rwE   
...                       ...                   ...                   ...   
1024609  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV-O48jNFNDElBu   
1024610  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV-O48jNFNDElBv   
1024611  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV_O48jNFNDElBw   
1024612  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV-O48jNFNDElBt   
1024613  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV_O48jNFNDElBx   

                  TYPE                    RULE SEVERITY  STATUS RESOLUTION  \
0           CODE_SMELL             squid:S1133     INFO    OPEN        NaN   
1           CODE_SMELL             squid:S1133     INFO    OPEN        NaN   
2           CODE_SMELL             squid:S1133     INFO    OPEN        NaN   
3           CODE_SMELL             squid:S1133     INFO    OPEN        NaN   
4           CODE_SMELL             squid:S1135     INFO  CLOSED      FIXED   
...                ...                     ...      ...     ...        ...   
1024609     CODE_SMELL            squid:S00117    MINOR    OPEN        NaN   
1024610            BUG             squid:S1143  BLOCKER    OPEN        NaN   
1024611     CODE_SMELL             squid:S1161    MAJOR    OPEN        NaN   
1024612     CODE_SMELL  squid:HiddenFieldCheck    MAJOR    OPEN        NaN   
1024613  VULNERABILITY             squid:S1313    MAJOR    OPEN        NaN   

         EFFORT  DEBT  ...                                            MESSAGE  \
0          10.0  10.0  ...  Do not forget to remove this deprecated code s...   
1          10.0  10.0  ...  Do not forget to remove this deprecated code s...   
2          10.0  10.0  ...  Do not forget to remove this deprecated code s...   
3          10.0  10.0  ...  Do not forget to remove this deprecated code s...   
4           NaN   NaN  ...  Complete the task associated to this TODO comm...   
...         ...   ...  ...                                                ...   
1024609     2.0   2.0  ...  Rename this local variable to match the regula...   
1024610    30.0  30.0  ...  Remove this return statement from this finally...   
1024611     5.0   5.0  ...  Add the "@Override" annotation above this meth...   
1024612     5.0   5.0  ...  Rename "PORT" which hides the field declared a...   
1024613    30.0  30.0  ...     Make this IP "127.0.0.1" address configurable.   

                                                 COMPONENT START_LINE  \
0        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...       39.0   
1        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...       45.0   
2        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...       51.0   
3        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...       59.0   
4        Apache_Cayenne:framework/cayenne-jdk1.4-unpubl...       33.0   
...                                                    ...        ...   
1024609  org.apache:zookeeper:src/java/test/org/apache/...      202.0   
1024610  org.apache:zookeeper:src/java/test/org/apache/...      157.0   
1024611  org.apache:zookeeper:src/java/test/org/apache/...      206.0   
1024612  org.apache:zookeeper:src/java/test/org/apache/...      202.0   
1024613  org.apache:zookeeper:src/java/test/org/apache/...       47.0   

        END_LINE START_OFFSET  END_OFFSET  HASH  FROM_HOTSPOT  NOT_FOUND  \
0           39.0         27.0        37.0   NaN           NaN          0   
1           45.0         27.0        38.0   NaN           NaN          0   
2           51.0         24.0        37.0   NaN           NaN          0   
3  

## Filter out only Code Smell Issues

In [5]:
df1 = df1[df1['type'] == 'CODE_SMELL']

# CHANGE
df1 = df1[df1['status'] == 'OPEN']
df1

projectID          creationDate closeDate  \
0        commons-daemon  2003-09-04T23:28:19Z       NaN   
13       commons-daemon  2003-09-04T23:28:19Z       NaN   
18       commons-daemon  2003-09-04T23:28:19Z       NaN   
20       commons-daemon  2003-09-04T23:28:19Z       NaN   
21       commons-daemon  2003-09-04T23:28:19Z       NaN   
...                 ...                   ...       ...   
1941503     commons-cli  2017-07-29T14:36:30Z       NaN   
1941504     commons-cli  2017-07-29T14:36:30Z       NaN   
1941505     commons-cli  2017-07-29T14:36:30Z       NaN   
1941506     commons-cli  2018-02-26T17:23:40Z       NaN   
1941507     commons-cli  2018-02-26T17:23:40Z       NaN   

                               creationCommitHash closeCommitHash        type  \
0        d3416d3a25b16da3d18b3849522fa96183918e5b             NaN  CODE_SMELL   
13       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN  CODE_SMELL   
18       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN  CODE_SMELL   
20       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN  CODE_SMELL   
21       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN  CODE_SMELL   
...                                           ...             ...         ...   
1941503  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN  CODE_SMELL   
1941504  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN  CODE_SMELL   
1941505  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN  CODE_SMELL   
1941506  b0024d482050a08efc36c3cabee37c0af0e57a10             NaN  CODE_SMELL   
1941507  b0024d482050a08efc36c3cabee37c0af0e57a10             NaN  CODE_SMELL   

                squid                                          component  \
0        squid:S00112  org.apache:daemon:src/main/java/org/apache/com...   
13        squid:S1118  org.apache:daemon:src/main/java/org/apache/com...   
18       squid:S00122  org.apache:daemon:src/samples/ServiceDaemonRea...   
20       squid:S00122    org.apache:daemon:src/samples/SimpleDaemon.java   
21       squid:S00122    org.apache:daemon:src/samples/SimpleDaemon.java   
...               ...                                                ...   
1941503   squid:S1192  org.apache:commons-cli:src/test/java/org/apach...   
1941504   squid:S1192  org.apache:commons-cli:src/test/java/org/apach...   
1941505   squid:S1192  org.apache:commons-cli:src/test/java/org/apach...   
1941506  squid:S00112  org.apache:commons-cli:src/test/java/org/apach...   
1941507  squid:S00100  org.apache:commons-cli:src/test/java/org/apach...   

        severity                 project  startLine  endLine resolution  \
0          MAJOR       org.apache:daemon       71.0     71.0        NaN   
13         MAJOR       org.apache:daemon       30.0     30.0        NaN   
18         MINOR       org.apache:daemon       34.0     34.0        NaN   
20         MINOR       org.apache:daemon       78.0     78.0        NaN   
21         MINOR       org.apache:daemon      132.0    132.0        NaN   
...          ...                     ...        ...      ...        ...   
1941503    MINOR  org.apache:commons-cli       34.0     34.0        NaN   
1941504    MINOR  org.apache:commons-cli       36.0     36.0        NaN   
1941505    MINOR  org.apache:commons-cli       35.0     35.0        NaN   
1941506    MAJOR  org.apache:commons-cli      153.0    153.0        NaN   
1941507    MINOR  org.apache:commons-cli      152.0    152.0        NaN   

        status                                            message  effort  \
0         OPEN  Define and throw a dedicated exception instead...      20   
13        OPEN  Add a private constructor to hide the implicit...      30   
18        OPEN  At most one statement is allowed per line, but...       1   
20        OPEN  At most one statement is allowed per line, but...       1   
21        OPEN  At most one statement is allowed per line, but...       1   
...        ...                                          

In [6]:
df2 = df2[df2['TYPE'] == 'CODE_SMELL']

# CHANGE
df2 = df2[df2['STATUS'] == 'OPEN']
df2

PROJECT_ID CREATION_ANALYSIS_KEY             ISSUE_KEY  \
0          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWeF1RcPC4KKKThcBEZ8   
1          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWeF1RcRC4KKKThcBEZ9   
2          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWeF1RcSC4KKKThcBEZ-   
3          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWeF1RcSC4KKKThcBEZ_   
5          org.apache:cayenne  AWd5_psxC4KKKThc-qK6  AWd5_sHnC4KKKThc-sYd   
...                       ...                   ...                   ...   
1024607  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV6O48jNFNDElBr   
1024608  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV-O48jNFNDElBs   
1024609  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV-O48jNFNDElBu   
1024611  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV_O48jNFNDElBw   
1024612  org.apache:zookeeper  AWMdqqQTO48jNFNDElBo  AWMdqrV-O48jNFNDElBt   

               TYPE                            RULE SEVERITY STATUS  \
0        CODE_SMELL                     squid:S1133     INFO   OPEN   
1        CODE_SMELL                     squid:S1133     INFO   OPEN   
2        CODE_SMELL                     squid:S1133     INFO   OPEN   
3        CODE_SMELL                     squid:S1133     INFO   OPEN   
5        CODE_SMELL                     squid:S1135     INFO   OPEN   
...             ...                             ...      ...    ...   
1024607  CODE_SMELL    common-java:DuplicatedBlocks    MAJOR   OPEN   
1024608  CODE_SMELL  squid:EmptyStatementUsageCheck    MINOR   OPEN   
1024609  CODE_SMELL                    squid:S00117    MINOR   OPEN   
1024611  CODE_SMELL                     squid:S1161    MAJOR   OPEN   
1024612  CODE_SMELL          squid:HiddenFieldCheck    MAJOR   OPEN   

        RESOLUTION  EFFORT  DEBT  ...  \
0              NaN    10.0  10.0  ...   
1              NaN    10.0  10.0  ...   
2              NaN    10.0  10.0  ...   
3              NaN    10.0  10.0  ...   
5              NaN     NaN   NaN  ...   
...            ...     ...   ...  ...   
1024607        NaN    30.0  30.0  ...   
1024608        NaN     2.0   2.0  ...   
1024609        NaN     2.0   2.0  ...   
1024611        NaN     5.0   5.0  ...   
1024612        NaN     5.0   5.0  ...   

                                                   MESSAGE  \
0        Do not forget to remove this deprecated code s...   
1        Do not forget to remove this deprecated code s...   
2        Do not forget to remove this deprecated code s...   
3        Do not forget to remove this deprecated code s...   
5        Complete the task associated to this TODO comm...   
...                                                    ...   
1024607       2 duplicated blocks of code must be removed.   
1024608                       Remove this empty statement.   
1024609  Rename this local variable to match the regula...   
1024611  Add the "@Override" annotation above this meth...   
1024612  Rename "PORT" which hides the field declared a...   

                                                 COMPONENT START_LINE  \
0        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...       39.0   
1        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...       45.0   
2        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...       51.0   
3        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...       59.0   
5        Apache_Cayenne:framework/cayenne-jdk1.5-unpubl...      172.0   
...                                                    ...        ...   
1024607  org.apache:zookeeper:src/java/test/org/apache/...        NaN   
1024608  org.apache:zookeeper:src/java/test/org/apache/...      218.0   
1024609  org.apache:zookeeper:src/java/test/org/apache/...      202.0   
1024611  org.apache:zookeeper:src/java/test/org/apache/...      206.0   
1024612  org.apache:zookeeper:src/java/test/org/apache/...      202.0   

        END_LINE START_OFFSET  END_OFFSET  HASH  FROM_HOTSPOT  NOT_FOUND  \
0           39.0         27.0        37.0   NaN   

## Select only the necessary columns
To save space, columns that do not directly contribute to identifying the projects and commits / analysis or helps with bringing the two database versions together are removed.

In [7]:
df1.columns

Index(['projectID', 'creationDate', 'closeDate', 'creationCommitHash',
       'closeCommitHash', 'type', 'squid', 'component', 'severity', 'project',
       'startLine', 'endLine', 'resolution', 'status', 'message', 'effort',
       'debt', 'author'],
      dtype='object')

In [8]:
df1 = df1[["projectID", "type", "creationDate", "creationCommitHash", "closeCommitHash", "status", "message"]]
df1

projectID        type          creationDate  \
0        commons-daemon  CODE_SMELL  2003-09-04T23:28:19Z   
13       commons-daemon  CODE_SMELL  2003-09-04T23:28:19Z   
18       commons-daemon  CODE_SMELL  2003-09-04T23:28:19Z   
20       commons-daemon  CODE_SMELL  2003-09-04T23:28:19Z   
21       commons-daemon  CODE_SMELL  2003-09-04T23:28:19Z   
...                 ...         ...                   ...   
1941503     commons-cli  CODE_SMELL  2017-07-29T14:36:30Z   
1941504     commons-cli  CODE_SMELL  2017-07-29T14:36:30Z   
1941505     commons-cli  CODE_SMELL  2017-07-29T14:36:30Z   
1941506     commons-cli  CODE_SMELL  2018-02-26T17:23:40Z   
1941507     commons-cli  CODE_SMELL  2018-02-26T17:23:40Z   

                               creationCommitHash closeCommitHash status  \
0        d3416d3a25b16da3d18b3849522fa96183918e5b             NaN   OPEN   
13       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN   OPEN   
18       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN   OPEN   
20       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN   OPEN   
21       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN   OPEN   
...                                           ...             ...    ...   
1941503  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN   OPEN   
1941504  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN   OPEN   
1941505  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN   OPEN   
1941506  b0024d482050a08efc36c3cabee37c0af0e57a10             NaN   OPEN   
1941507  b0024d482050a08efc36c3cabee37c0af0e57a10             NaN   OPEN   

                                                   message  
0        Define and throw a dedicated exception instead...  
13       Add a private constructor to hide the implicit...  
18       At most one statement is allowed per line, but...  
20       At most one statement is allowed per line, but...  
21       At most one statement is allowed per line, but...  
...                                                    ...  
1941503  Define a constant instead of duplicating this ...  
1941504  Define a constant instead of duplicating this ...  
1941505  Define a constant instead of duplicating this ...  
1941506  Define and throw a dedicated exception instead...  
1941507  Rename this method name to match the regular e...  

[119254 rows x 7 columns]

In [9]:
df2.columns

Index(['PROJECT_ID', 'CREATION_ANALYSIS_KEY', 'ISSUE_KEY', 'TYPE', 'RULE',
       'SEVERITY', 'STATUS', 'RESOLUTION', 'EFFORT', 'DEBT', 'TAGS',
       'CREATION_DATE', 'CLOSE_DATE', 'MESSAGE', 'COMPONENT', 'START_LINE',
       'END_LINE', 'START_OFFSET', 'END_OFFSET', 'HASH', 'FROM_HOTSPOT',
       'NOT_FOUND', 'CLOSE_ANALYSIS_KEY'],
      dtype='object')

In [10]:
df2 = df2[["PROJECT_ID", "TYPE", "TAGS", "CREATION_DATE", "CREATION_ANALYSIS_KEY", "STATUS", "MESSAGE"]]
df2

PROJECT_ID        TYPE               TAGS  \
0          org.apache:cayenne  CODE_SMELL           obsolete   
1          org.apache:cayenne  CODE_SMELL           obsolete   
2          org.apache:cayenne  CODE_SMELL           obsolete   
3          org.apache:cayenne  CODE_SMELL           obsolete   
5          org.apache:cayenne  CODE_SMELL                cwe   
...                       ...         ...                ...   
1024607  org.apache:zookeeper  CODE_SMELL            pitfall   
1024608  org.apache:zookeeper  CODE_SMELL  cert,misra,unused   
1024609  org.apache:zookeeper  CODE_SMELL         convention   
1024611  org.apache:zookeeper  CODE_SMELL       bad-practice   
1024612  org.apache:zookeeper  CODE_SMELL       cert,pitfall   

               CREATION_DATE CREATION_ANALYSIS_KEY STATUS  \
0        2007-01-21 11:11:02  AWd5_psxC4KKKThc-qK6   OPEN   
1        2007-01-21 11:11:02  AWd5_psxC4KKKThc-qK6   OPEN   
2        2007-01-21 11:11:02  AWd5_psxC4KKKThc-qK6   OPEN   
3        2007-01-21 11:11:02  AWd5_psxC4KKKThc-qK6   OPEN   
5        2007-01-21 11:11:02  AWd5_psxC4KKKThc-qK6   OPEN   
...                      ...                   ...    ...   
1024607  2016-12-21 22:32:01  AWMdqqQTO48jNFNDElBo   OPEN   
1024608  2016-12-21 22:32:01  AWMdqqQTO48jNFNDElBo   OPEN   
1024609  2016-12-21 22:32:01  AWMdqqQTO48jNFNDElBo   OPEN   
1024611  2016-12-21 22:32:01  AWMdqqQTO48jNFNDElBo   OPEN   
1024612  2016-12-21 22:32:01  AWMdqqQTO48jNFNDElBo   OPEN   

                                                   MESSAGE  
0        Do not forget to remove this deprecated code s...  
1        Do not forget to remove this deprecated code s...  
2        Do not forget to remove this deprecated code s...  
3        Do not forget to remove this deprecated code s...  
5        Complete the task associated to this TODO comm...  
...                                                    ...  
1024607       2 duplicated blocks of code must be removed.  
1024608                       Remove this empty statement.  
1024609  Rename this local variable to match the regula...  
1024611  Add the "@Override" annotation above this meth...  
1024612  Rename "PORT" which hides the field declared a...  

[216190 rows x 7 columns]

## Try to bring together the two database versions
Version 1 doesn't contain any tags in the issue table. Both versions do contain a message though. It is to be investigated whether tags for version 1 can be created from the messages. Assumption: If the tags are consistent with the messages, a certain issue message makes it possible to construct tags for version 1.

## Check consistency of tags with messages for database version 2

In [11]:
check = df2[["TAGS", "MESSAGE"]]
check

TAGS                                            MESSAGE
0                 obsolete  Do not forget to remove this deprecated code s...
1                 obsolete  Do not forget to remove this deprecated code s...
2                 obsolete  Do not forget to remove this deprecated code s...
3                 obsolete  Do not forget to remove this deprecated code s...
5                      cwe  Complete the task associated to this TODO comm...
...                    ...                                                ...
1024607            pitfall       2 duplicated blocks of code must be removed.
1024608  cert,misra,unused                       Remove this empty statement.
1024609         convention  Rename this local variable to match the regula...
1024611       bad-practice  Add the "@Override" annotation above this meth...
1024612       cert,pitfall  Rename "PORT" which hides the field declared a...

[216190 rows x 2 columns]

In [12]:
check['PROCESSED_MESSAGE'] = df2['MESSAGE'].copy()
check

C:\Users\carol\AppData\Local\Temp\ipykernel_15268\3609162278.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  check['PROCESSED_MESSAGE'] = df2['MESSAGE'].copy()


TAGS                                            MESSAGE  \
0                 obsolete  Do not forget to remove this deprecated code s...   
1                 obsolete  Do not forget to remove this deprecated code s...   
2                 obsolete  Do not forget to remove this deprecated code s...   
3                 obsolete  Do not forget to remove this deprecated code s...   
5                      cwe  Complete the task associated to this TODO comm...   
...                    ...                                                ...   
1024607            pitfall       2 duplicated blocks of code must be removed.   
1024608  cert,misra,unused                       Remove this empty statement.   
1024609         convention  Rename this local variable to match the regula...   
1024611       bad-practice  Add the "@Override" annotation above this meth...   
1024612       cert,pitfall  Rename "PORT" which hides the field declared a...   

                                         PROCESSED_MESSAGE  
0        Do not forget to remove this deprecated code s...  
1        Do not forget to remove this deprecated code s...  
2        Do not forget to remove this deprecated code s...  
3        Do not forget to remove this deprecated code s...  
5        Complete the task associated to this TODO comm...  
...                                                    ...  
1024607       2 duplicated blocks of code must be removed.  
1024608                       Remove this empty statement.  
1024609  Rename this local variable to match the regula...  
1024611  Add the "@Override" annotation above this meth...  
1024612  Rename "PORT" which hides the field declared a...  

[216190 rows x 3 columns]

In [13]:
# Are messages available for all found code smells? - Yes.
check[check['MESSAGE'].isna()]

Empty DataFrame
Columns: [TAGS, MESSAGE, PROCESSED_MESSAGE]
Index: []

### Make the messages comparable
The issue messages contain direct names and numbers to describe variables or lines directly. This leads to messages having a similar goal seeming unique due to the specifics. Names and numbers need to be stripped, so the messages become comparable over the whole column.

In [14]:
def process_string(text: str) -> str:
    """ This function removes variable names that are recognizable by being encased in " " or ' '. """
    if isinstance(text, str):
        # remove substrings between double or single quotes
        processed_text = re.sub(r"(this literal).*", r"\1", text)
        processed_text = re.sub(r'"[^"]*"', '', processed_text)
        processed_text = re.sub(r"'[^']*'", '', processed_text)
        return processed_text.strip()

check['PROCESSED_MESSAGE'] = check['PROCESSED_MESSAGE'].apply(process_string)
check

C:\Users\carol\AppData\Local\Temp\ipykernel_15268\583907064.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  check['PROCESSED_MESSAGE'] = check['PROCESSED_MESSAGE'].apply(process_string)


TAGS                                            MESSAGE  \
0                 obsolete  Do not forget to remove this deprecated code s...   
1                 obsolete  Do not forget to remove this deprecated code s...   
2                 obsolete  Do not forget to remove this deprecated code s...   
3                 obsolete  Do not forget to remove this deprecated code s...   
5                      cwe  Complete the task associated to this TODO comm...   
...                    ...                                                ...   
1024607            pitfall       2 duplicated blocks of code must be removed.   
1024608  cert,misra,unused                       Remove this empty statement.   
1024609         convention  Rename this local variable to match the regula...   
1024611       bad-practice  Add the "@Override" annotation above this meth...   
1024612       cert,pitfall  Rename "PORT" which hides the field declared a...   

                                         PROCESSED_MESSAGE  
0        Do not forget to remove this deprecated code s...  
1        Do not forget to remove this deprecated code s...  
2        Do not forget to remove this deprecated code s...  
3        Do not forget to remove this deprecated code s...  
5        Complete the task associated to this TODO comm...  
...                                                    ...  
1024607       2 duplicated blocks of code must be removed.  
1024608                       Remove this empty statement.  
1024609  Rename this local variable to match the regula...  
1024611    Add the  annotation above this method signature  
1024612  Rename  which hides the field declared at line...  

[216190 rows x 3 columns]

In [15]:
# remove numbers from strings
check['PROCESSED_MESSAGE'] = check['PROCESSED_MESSAGE'].str.replace(r'\d+', '', regex=True)
check

C:\Users\carol\AppData\Local\Temp\ipykernel_15268\2890904256.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  check['PROCESSED_MESSAGE'] = check['PROCESSED_MESSAGE'].str.replace(r'\d+', '', regex=True)


TAGS                                            MESSAGE  \
0                 obsolete  Do not forget to remove this deprecated code s...   
1                 obsolete  Do not forget to remove this deprecated code s...   
2                 obsolete  Do not forget to remove this deprecated code s...   
3                 obsolete  Do not forget to remove this deprecated code s...   
5                      cwe  Complete the task associated to this TODO comm...   
...                    ...                                                ...   
1024607            pitfall       2 duplicated blocks of code must be removed.   
1024608  cert,misra,unused                       Remove this empty statement.   
1024609         convention  Rename this local variable to match the regula...   
1024611       bad-practice  Add the "@Override" annotation above this meth...   
1024612       cert,pitfall  Rename "PORT" which hides the field declared a...   

                                         PROCESSED_MESSAGE  
0        Do not forget to remove this deprecated code s...  
1        Do not forget to remove this deprecated code s...  
2        Do not forget to remove this deprecated code s...  
3        Do not forget to remove this deprecated code s...  
5        Complete the task associated to this TODO comm...  
...                                                    ...  
1024607         duplicated blocks of code must be removed.  
1024608                       Remove this empty statement.  
1024609  Rename this local variable to match the regula...  
1024611    Add the  annotation above this method signature  
1024612   Rename  which hides the field declared at line .  

[216190 rows x 3 columns]

In [16]:
check["PROCESSED_MESSAGE"].unique().tolist()[:100]

['Do not forget to remove this deprecated code someday.',
 'Complete the task associated to this TODO comment.',
 'Move this constructor to comply with Java Code Conventions.',
 'Complex class',
 'Make this line start at column .',
 'Tab characters found (this is the first occurrence)',
 'Replace all tab characters in this file by sequences of white-spaces.',
 'Rename this class name to match the regular expression .',
 'Define a constant instead of duplicating this literal',
 'Define this declaration on a separate line.',
 'Rename selector  to match the regular expression: ^[a-z][a-z\\-]*$',
 'Move the array designator from the variable to the type.',
 'Make the  logger private static final and rename it to comply with the format .',
 'Remove the declaration of thrown exception  which is a runtime exception.',
 'Rename this local variable to match the regular expression .',
 'Move constants to a class or enum.',
 'Remove the redundant  thrown exception declaration(s).',
 'Directly app

In [17]:
unique_messages = check["PROCESSED_MESSAGE"].unique().tolist()
len(unique_messages)

328

In [18]:
# check whether removing duplicates from the df with only tags and processed_message has the same lengthv
check = check.drop_duplicates(subset=['TAGS', 'PROCESSED_MESSAGE'])
check

TAGS  \
0                      obsolete   
5                           cwe   
503                  convention   
517      antipattern,code-smell   
560                  convention   
...                         ...   
1019542          error-handling   
1021194          error-handling   
1021318          error-handling   
1022791          error-handling   
1024337          error-handling   

                                                   MESSAGE  \
0        Do not forget to remove this deprecated code s...   
5        Complete the task associated to this TODO comm...   
503      Move this constructor to comply with Java Code...   
517                                          Complex class   
560                      Make this line start at column 9.   
...                                                    ...   
1019542  Refactor this method to throw at most one chec...   
1021194  Refactor this method to throw at most one chec...   
1021318  Refactor this method to throw at most one chec...   
1022791  Refactor this method to throw at most one chec...   
1024337  Refactor this method to throw at most one chec...   

                                         PROCESSED_MESSAGE  
0        Do not forget to remove this deprecated code s...  
5        Complete the task associated to this TODO comm...  
503      Move this constructor to comply with Java Code...  
517                                          Complex class  
560                       Make this line start at column .  
...                                                    ...  
1019542  Refactor this method to throw at most one chec...  
1021194  Refactor this method to throw at most one chec...  
1021318  Refactor this method to throw at most one chec...  
1022791  Refactor this method to throw at most one chec...  
1024337  Refactor this method to throw at most one chec...  

[329 rows x 3 columns]

In [19]:
# Which processed_messages has a different tag?

# group by 'PROCESSED_MESSAGE'
grouped = check.groupby('PROCESSED_MESSAGE')

# filter for groups where the number of unique TAGS sets is greater than 1
messages_with_different_tags = grouped.filter(lambda x: x['TAGS'].nunique() > 1)

messages_with_different_tags

TAGS                                            MESSAGE  \
671           convention  Replace all tab characters in this file by seq...   
1012738  convention,psr2  Replace all tab characters in this file by seq...   

                                         PROCESSED_MESSAGE  
671      Replace all tab characters in this file by seq...  
1012738  Replace all tab characters in this file by seq...

### Comparability: Result
Since the length of a dataframe containing only unique combinations of TAGS and PROCESSED_MESSAGE compared to the length of unique processed messages, it's worth trying to infer tags for known messages for version 1.

## Investigating Tags further
It is important to investigate how many tags there are, how many are missing and what kinds of tags exist in the data.

In [20]:
# How many rows have no tag(s)? - 3.
check[check["TAGS"].isna()]

TAGS                                            MESSAGE  \
817    NaN                 Missing newline after last element   
1557   NaN            Start every element on a separate line.   
12854  NaN  The Cyclomatic Complexity of this class is 407...   

                                       PROCESSED_MESSAGE  
817                   Missing newline after last element  
1557             Start every element on a separate line.  
12854  The Cyclomatic Complexity of this class is  wh...

In [21]:
# How many unique tags are there? - 60, inluding combinations.
print(len(check["TAGS"].unique().tolist()))
print(check["TAGS"].unique().tolist())

60
['obsolete', 'cwe', 'convention', 'antipattern,code-smell', 'design', 'format', 'clumsy,redundant,unused', 'bad-practice', 'clumsy', 'clumsy,redundant', nan, 'serialization,unused', 'style', 'bad-practice,misra', 'cert,misra,unused', 'unused', 'bad-practice,obsolete', 'performance', 'pitfall', 'bad-practice,cert', 'cert', 'brain-overload', 'java8', 'misra,pitfall', 'obsolete,pitfall', 'design,performance', 'convention,performance', 'convention,pitfall', 'browser-compatibility,convention', 'html5,obsolete', 'cert,pitfall', 'suspicious', 'cert,cwe,error-handling', 'misra,unused', 'cert,cwe,misra', 'error-handling', 'confusing', 'cert,cwe,misra,suspicious', 'cert,unused', 'cert,clumsy,error-handling', 'cert,misra,pitfall', 'cert,error-handling,suspicious', 'cert,multi-threading', 'cert,cwe,suspicious', 'misra,suspicious', 'bad-practice,unpredictable', 'bad-practice,cert,cwe,error-handling', 'performance,pitfall,sql', 'design,pitfall', 'lock-in,pitfall', 'browser-compatibility', 'perfor

In [22]:
# How many unique tags are there without combinations? - 31.
tag_list = check["TAGS"].unique().tolist()
unique_items = set()

for item in tag_list:
    # check whether item from list is a string
    if isinstance(item, str):
        if ',' in item:
            sub_items = item.split(',')
            for sub_item in sub_items:
                # remove whitespace
                unique_items.add(sub_item.strip())
        else:
            unique_items.add(item.strip())

print(f"The unique items in the list are: {sorted(list(unique_items))}")
print(len(sorted(list(unique_items))))

The unique items in the list are: ['antipattern', 'bad-practice', 'brain-overload', 'browser-compatibility', 'cert', 'clumsy', 'code-smell', 'confusing', 'convention', 'cross-browser', 'cwe', 'design', 'error-handling', 'format', 'html5', 'java8', 'lock-in', 'misra', 'multi-threading', 'obsolete', 'performance', 'pitfall', 'psr2', 'redundant', 'serialization', 'sql', 'style', 'suspicious', 'unpredictable', 'unused', 'user-experience']
31


## Compare messages of the both versions
The goal is to infer the tags for version 1 from the known messages of version 2. It's also important to see how many of the tags have successfully been infered out of all the data there is. For this, the following steps are needed:
* Transform the messages according to version 2, so they can be generalised.
* Compare the transformed messages of version 1 to the known messages from version 2, to see how well they align.
* If there are a lot of common messages, infer the tags.

### Transform the messages of version 2

In [23]:
check2 = pd.DataFrame(df1["message"])
check2["processedMessage"] = df1["message"].copy()
check2

message  \
0        Define and throw a dedicated exception instead...   
13       Add a private constructor to hide the implicit...   
18       At most one statement is allowed per line, but...   
20       At most one statement is allowed per line, but...   
21       At most one statement is allowed per line, but...   
...                                                    ...   
1941503  Define a constant instead of duplicating this ...   
1941504  Define a constant instead of duplicating this ...   
1941505  Define a constant instead of duplicating this ...   
1941506  Define and throw a dedicated exception instead...   
1941507  Rename this method name to match the regular e...   

                                          processedMessage  
0        Define and throw a dedicated exception instead...  
13       Add a private constructor to hide the implicit...  
18       At most one statement is allowed per line, but...  
20       At most one statement is allowed per line, but...  
21       At most one statement is allowed per line, but...  
...                                                    ...  
1941503  Define a constant instead of duplicating this ...  
1941504  Define a constant instead of duplicating this ...  
1941505  Define a constant instead of duplicating this ...  
1941506  Define and throw a dedicated exception instead...  
1941507  Rename this method name to match the regular e...  

[119254 rows x 2 columns]

In [24]:
# apply transformations
check2['processedMessage'] = check2['processedMessage'].apply(process_string)
check2['processedMessage'] = check2['processedMessage'].str.replace(r'\d+', '', regex=True)
check2

message  \
0        Define and throw a dedicated exception instead...   
13       Add a private constructor to hide the implicit...   
18       At most one statement is allowed per line, but...   
20       At most one statement is allowed per line, but...   
21       At most one statement is allowed per line, but...   
...                                                    ...   
1941503  Define a constant instead of duplicating this ...   
1941504  Define a constant instead of duplicating this ...   
1941505  Define a constant instead of duplicating this ...   
1941506  Define and throw a dedicated exception instead...   
1941507  Rename this method name to match the regular e...   

                                          processedMessage  
0        Define and throw a dedicated exception instead...  
13       Add a private constructor to hide the implicit...  
18       At most one statement is allowed per line, but...  
20       At most one statement is allowed per line, but...  
21       At most one statement is allowed per line, but...  
...                                                    ...  
1941503  Define a constant instead of duplicating this ...  
1941504  Define a constant instead of duplicating this ...  
1941505  Define a constant instead of duplicating this ...  
1941506  Define and throw a dedicated exception instead...  
1941507  Rename this method name to match the regular e...  

[119254 rows x 2 columns]

### Comparing the messages of version 1 and 2

In [25]:
# How many unique messages are there?
len1 = len(check2["processedMessage"].unique().tolist())
len2 = len(check["PROCESSED_MESSAGE"].unique().tolist())
print(f"Version 1: {len1}")
print(f"Version 2: {len2}")

Version 1: 263
Version 2: 328


In [26]:
set1 = set(list(check["PROCESSED_MESSAGE"]))
set2 = set(list(check2["processedMessage"]))

common_elements = list(set1.intersection(set2))
common_elements

['Remove this unnecessary import: java.lang classes are always implicitly imported.',
 'Refactor this method to throw at most one checked exception instead of: java.lang.Exception, java.lang.InterruptedException',
 'Use  for this string-to-float conversion.',
 'Refactor this method to throw at most one checked exception instead of: javax.xml.parsers.ParserConfigurationException, java.io.IOException, org.xml.sax.SAXException',
 'Refactor this method to throw at most one checked exception instead of: java.security.NoSuchProviderException, java.security.NoSuchAlgorithmException',
 'Replace all tab characters in this file by sequences of white-spaces.',
 'Remove these useless parentheses.',
 'Refused parent bequest',
 'Introduce a new variable instead of reusing the parameter .',
 'Rename this constant name to match the regular expression .',
 'Refactor this method to throw at most one checked exception instead of: javax.management.ReflectionException, javax.management.MBeanException',
 'R

In [27]:
len(common_elements)

209

Version 1 has 263 different issue messages and version 2 has 328. They only share 209, so for these messages only there can be tags infered from version 2. <br>
Which messages of version 1 are not shared by version 2?

In [28]:
exclusive_in_set2 = list(set2.difference(set1))
exclusive_in_set2

['V is not used in the method.',
 'InputT is not used in the interface.',
 'Add a  method to this class.',
 'Refactor this method to throw at most one checked exception instead of: java.security.NoSuchAlgorithmException, javax.crypto.NoSuchPaddingException, java.security.NoSuchProviderException, java.security.InvalidKeyException, javax.crypto.IllegalBlockSizeException, javax.crypto.BadPaddingException',
 'OutputT is not used in the interface.',
 'AccumT is not used in the class.',
 'Refactor this method to throw at most one checked exception instead of: javax.xml.parsers.ParserConfigurationException, org.xml.sax.SAXException, java.io.IOException, javax.xml.transform.TransformerException, javax.xml.xpath.XPathExpressionException',
 'Remove useless curly braces around statement (sonar.java.source not set. Assuming  or greater.)',
 'Refactor this method to throw at most one checked exception instead of: java.net.URISyntaxException, java.net.MalformedURLException',
 'Move the id\\\\id\\\\ 

A lot of these messages contain similar parts. So they are checked one-by-one so their tags might be correctly infered even though they aren't in version 2 directly.

In [29]:
check[check['PROCESSED_MESSAGE'].str.contains("is not used in the method.")]

TAGS                       MESSAGE  \
1005536  cert,unused  N is not used in the method.   

                    PROCESSED_MESSAGE  
1005536  N is not used in the method.

In [30]:
pd.set_option('display.max_rows', None)
check[check['PROCESSED_MESSAGE'].str.contains("Refactor this method to throw at most one checked exception instead of:")]

TAGS                                            MESSAGE  \
9754     error-handling  Refactor this method to throw at most one chec...   
9950     error-handling  Refactor this method to throw at most one chec...   
11828    error-handling  Refactor this method to throw at most one chec...   
15573    error-handling  Refactor this method to throw at most one chec...   
23287    error-handling  Refactor this method to throw at most one chec...   
102933   error-handling  Refactor this method to throw at most one chec...   
102960   error-handling  Refactor this method to throw at most one chec...   
102963   error-handling  Refactor this method to throw at most one chec...   
102965   error-handling  Refactor this method to throw at most one chec...   
107535   error-handling  Refactor this method to throw at most one chec...   
135802   error-handling  Refactor this method to throw at most one chec...   
135805   error-handling  Refactor this method to throw at most one chec...   
135806   error-handling  Refactor this method to throw at most one chec...   
136551   error-handling  Refactor this method to throw at most one chec...   
136606   error-handling  Refactor this method to throw at most one chec...   
138162   error-handling  Refactor this method to throw at most one chec...   
139550   error-handling  Refactor this method to throw at most one chec...   
151719   error-handling  Refactor this method to throw at most one chec...   
191924   error-handling  Refactor this method to throw at most one chec...   
209584   error-handling  Refactor this method to throw at most one chec...   
267083   error-handling  Refactor this method to throw at most one chec...   
270935   error-handling  Refactor this method to throw at most one chec...   
314467   error-handling  Refactor this method to throw at most one chec...   
333900   error-handling  Refactor this method to throw at most one chec...   
333901   error-handling  Refactor this method to throw at most one chec...   
334011   error-handling  Refactor this method to throw at most one chec...   
334039   error-handling  Refactor this method to throw at most one chec...   
334988   error-handling  Refactor this method to throw at most one chec...   
338027   error-handling  Refactor this method to throw at most one chec...   
338044   error-handling  Refactor this method to throw at most one chec...   
340120   error-handling  Refactor this method to throw at most one chec...   
340123   error-handling  Refactor this method to throw at most one chec...   
340251   error-handling  Refactor this method to throw at most one chec...   
340284   error-handling  Refactor this method to throw at most one chec...   
340332   error-handling  Refactor this method to throw at most one chec...   
340333   error-handling  Refactor this method to throw at most one chec...   
340335   error-handling  Refactor this method to throw at most one chec...   
340376   error-handling  Refactor this method to throw at most one chec...   
340377   error-handling  Refactor this method to throw at most one chec...   
340399   error-handling  Refactor this method to throw at most one chec...   
340400   error-handling  Refactor this method to throw at most one chec...   
340406   error-handling  Refactor this method to throw at most one chec...   
341167   error-handling  Refactor this method to throw at most one chec...   
341362   error-handling  Refactor this method to throw at most one chec...   
341418   error-handling  Refactor this method to throw at most one chec...   
341430   error-handling  Refactor this method to throw at most one chec...   
341436   error-handling  Refactor this method to throw at most one chec...   
346264   error-handling  Refactor this method to throw at most one chec...   
369023   error-handling  Refactor this method to throw at most one chec...   
369024   error-handling  Refactor this method to throw at most one chec...   
371506   error-handling  Refactor this me

In [31]:
check[check['PROCESSED_MESSAGE'].str.contains("is not used in the class.")]

TAGS                      MESSAGE            PROCESSED_MESSAGE
398909   cert,unused  K is not used in the class.  K is not used in the class.
398910   cert,unused  V is not used in the class.  V is not used in the class.
456948   cert,unused  I is not used in the class.  I is not used in the class.
663089   cert,unused  T is not used in the class.  T is not used in the class.
1003813  cert,unused  E is not used in the class.  E is not used in the class.

In [32]:
check[check['PROCESSED_MESSAGE'].str.contains("literal on the left side of this string comparison.")]

TAGS                                            MESSAGE  \
10063   bad-practice  Move the "NSNumber" string literal on the left...   
348356  bad-practice  Move the "\"" string literal on the left side ...   
622919  bad-practice  Move the "\"java.lang.Enum\"" string literal o...   

                                        PROCESSED_MESSAGE  
10063   Move the  string literal on the left side of t...  
348356  Move the " string literal on the left side of ...  
622919  Move the java.lang.Enum\ string literal on the...

In [33]:
check[check['PROCESSED_MESSAGE'].str.contains("is not used in the interface.")]

TAGS                                 MESSAGE  \
847257  cert,unused  BatchKey is not used in the interface.   

                             PROCESSED_MESSAGE  
847257  BatchKey is not used in the interface.

In [34]:
check[check['PROCESSED_MESSAGE'].str.contains("Use super.clone")]

Empty DataFrame
Columns: [TAGS, MESSAGE, PROCESSED_MESSAGE]
Index: []

In [35]:
check[check['PROCESSED_MESSAGE'].str.contains("comply with the contract of the  method.")]

Empty DataFrame
Columns: [TAGS, MESSAGE, PROCESSED_MESSAGE]
Index: []

In [36]:
check[check['PROCESSED_MESSAGE'].str.contains("Remove the parentheses around the")]

Empty DataFrame
Columns: [TAGS, MESSAGE, PROCESSED_MESSAGE]
Index: []

In [37]:
check[check['PROCESSED_MESSAGE'].str.contains("Reduce this lambda expression number of lines")]

Empty DataFrame
Columns: [TAGS, MESSAGE, PROCESSED_MESSAGE]
Index: []

For all messages only contained in the version 1 data, there is either a consistent tag that it is associated with or no tag that can be infered.

### Check percentage coverage of messages from version 1 that can be infered

In [38]:
pd.reset_option('display.max_rows')

# create a variable to check whether tags can be inferred from version 2
def can_be_inferred(row):
    message = row['processedMessage']
    if message in common_elements:
        return True
    if "is not used in the method." in message:
        return True
    if "Refactor this method to throw at most one checked exception instead of:" in message:
        return True
    if "is not used in the class." in message:
        return True
    if "literal on the left side of this string comparison." in message:
        return True
    if "is not used in the interface." in message:
        return True
    return False


check2['canBeInferred'] = check2.apply(can_be_inferred, axis=1)
check2

message  \
0        Define and throw a dedicated exception instead...   
13       Add a private constructor to hide the implicit...   
18       At most one statement is allowed per line, but...   
20       At most one statement is allowed per line, but...   
21       At most one statement is allowed per line, but...   
...                                                    ...   
1941503  Define a constant instead of duplicating this ...   
1941504  Define a constant instead of duplicating this ...   
1941505  Define a constant instead of duplicating this ...   
1941506  Define and throw a dedicated exception instead...   
1941507  Rename this method name to match the regular e...   

                                          processedMessage  canBeInferred  
0        Define and throw a dedicated exception instead...           True  
13       Add a private constructor to hide the implicit...           True  
18       At most one statement is allowed per line, but...           True  
20       At most one statement is allowed per line, but...           True  
21       At most one statement is allowed per line, but...           True  
...                                                    ...            ...  
1941503  Define a constant instead of duplicating this ...           True  
1941504  Define a constant instead of duplicating this ...           True  
1941505  Define a constant instead of duplicating this ...           True  
1941506  Define and throw a dedicated exception instead...           True  
1941507  Rename this method name to match the regular e...           True  

[119254 rows x 3 columns]

In [39]:
# percentage of True values
print(f"Percentages of messages for which tags can be inferred: {round((check2['canBeInferred'].sum() / len(check2)) * 100, 2)} %")

Percentages of messages for which tags can be inferred: 99.92 %


### Result
Following the investigation, it can be concluded that almost all tags can be inferred from version 2, which increases the training data significantly.

In [40]:
# save check dfs
check.to_csv(os.path.join(data_dir, 'issues_v2.csv'), index = False)
check2.to_csv(os.path.join(data_dir, 'issues_v1.csv'), index = False)

## Prepare Version 2 data for tag prediction

### Flattening the labels over different analysis / commits
When pushing a commit and running the analysis, there can be multiple issues for one commit. For each of these commits, there can be multiple code smells. Each of these code smells can have multiple tags. There can be multiple code smells of the same type, so that in theory there would be a distribution of tags in comparison to all tags that SonarQube can assign to an issue. To make this predictable, the approach is to collect all tags for code smells for one commit and make this list the label. For this, the distribution of open (not fixed) code smells need to be flatted into a list.

In [41]:
# select columns that are needed to extract unique tags per analysis
df_tags = df2.groupby('CREATION_ANALYSIS_KEY')['TAGS'].unique().reset_index(name='UNIQUE_TAGS')
df_tags

CREATION_ANALYSIS_KEY                                        UNIQUE_TAGS
0     AV0-6-3jt6tne_r58pUF  [bad-practice,cert, cert,error-handling,suspic...
1     AV0-8Tovt6tne_r58pUY                                           [clumsy]
2     AV0-9p72t6tne_r58pUo                                     [bad-practice]
3     AV0-BQ2Et6tne_r58pCR                                           [design]
4     AV0-LoN2t6tne_r58pFZ                               [cert,pitfall, cert]
...                    ...                                                ...
9688  AWentw92Z1vVHhGU4u0K  [bad-practice,cert, confusing, cert,error-hand...
9689  AWenw12KZ1vVHhGU4vEw                              [pitfall, convention]
9690  AWenwUBsZ1vVHhGU4vAM    [convention, brain-overload, bad-practice,cert]
9691  AWenyFZ8Z1vVHhGU4vLr             [bad-practice,cert,cwe,error-handling]
9692  AWenyYdHZ1vVHhGU4vNW  [design, brain-overload, clumsy, unused, cert,...

[9693 rows x 2 columns]

This result shows the unique tag combinations associated with each analysis.

In [42]:
df_tags.loc[9688,"UNIQUE_TAGS"]

array(['bad-practice,cert', 'confusing', 'cert,error-handling,suspicious'],
      dtype=object)

As can be seen, the combinations lead to tags still appearing twice which needs to be further processed.

In [43]:
def safe_split(value):
    """Safely splits comma-separated strings within a NumPy array into unique tag values or returns an empty list."""
    if isinstance(value, np.ndarray):
        categories = []
        for item in value:
            if isinstance(item, str):
                categories.extend([cat.strip() for cat in item.split(',')])
        return list(set(categories))
    return []

df_tags["UNIQUE_TAGS"] = df_tags["UNIQUE_TAGS"].apply(safe_split)
df_tags

CREATION_ANALYSIS_KEY                                        UNIQUE_TAGS
0     AV0-6-3jt6tne_r58pUF   [error-handling, suspicious, bad-practice, cert]
1     AV0-8Tovt6tne_r58pUY                                           [clumsy]
2     AV0-9p72t6tne_r58pUo                                     [bad-practice]
3     AV0-BQ2Et6tne_r58pCR                                           [design]
4     AV0-LoN2t6tne_r58pFZ                                    [pitfall, cert]
...                    ...                                                ...
9688  AWentw92Z1vVHhGU4u0K  [error-handling, suspicious, confusing, bad-pr...
9689  AWenw12KZ1vVHhGU4vEw                              [convention, pitfall]
9690  AWenwUBsZ1vVHhGU4vAM   [cert, brain-overload, convention, bad-practice]
9691  AWenyFZ8Z1vVHhGU4vLr          [cwe, error-handling, bad-practice, cert]
9692  AWenyYdHZ1vVHhGU4vNW  [cwe, redundant, brain-overload, psr2, clumsy,...

[9693 rows x 2 columns]

In [44]:
df_tags.loc[9688,"UNIQUE_TAGS"]

['error-handling', 'suspicious', 'confusing', 'bad-practice', 'cert']

Now the strings are properly split and the list doesn't contain duplicate values.

In [45]:
df_tags.to_csv(os.path.join(data_dir, 'tags_analysiskey_v2.csv'), index = False)

## Generating tags for the version 1 dataset
Since it has been shown that it is worth to create tags by using the issue messages of version 1 and 2, there needs to be a kind of dictionary where it can be saved which tag goes to what message. For this the following is needed:
* A dataframe that maps the processed messages to the tags belonging to them
* A function that maps the tags to the messages of the version 1 data

### Dataframe that maps the processed messages to the tags

In [46]:
df_map = check[["TAGS", "PROCESSED_MESSAGE"]].drop_duplicates()
df_map

TAGS  \
0                      obsolete   
5                           cwe   
503                  convention   
517      antipattern,code-smell   
560                  convention   
...                         ...   
1019542          error-handling   
1021194          error-handling   
1021318          error-handling   
1022791          error-handling   
1024337          error-handling   

                                         PROCESSED_MESSAGE  
0        Do not forget to remove this deprecated code s...  
5        Complete the task associated to this TODO comm...  
503      Move this constructor to comply with Java Code...  
517                                          Complex class  
560                       Make this line start at column .  
...                                                    ...  
1019542  Refactor this method to throw at most one chec...  
1021194  Refactor this method to throw at most one chec...  
1021318  Refactor this method to throw at most one chec...  
1022791  Refactor this method to throw at most one chec...  
1024337  Refactor this method to throw at most one chec...  

[329 rows x 2 columns]

### Mapping the tags onto version 1

In [47]:
check2

message  \
0        Define and throw a dedicated exception instead...   
13       Add a private constructor to hide the implicit...   
18       At most one statement is allowed per line, but...   
20       At most one statement is allowed per line, but...   
21       At most one statement is allowed per line, but...   
...                                                    ...   
1941503  Define a constant instead of duplicating this ...   
1941504  Define a constant instead of duplicating this ...   
1941505  Define a constant instead of duplicating this ...   
1941506  Define and throw a dedicated exception instead...   
1941507  Rename this method name to match the regular e...   

                                          processedMessage  canBeInferred  
0        Define and throw a dedicated exception instead...           True  
13       Add a private constructor to hide the implicit...           True  
18       At most one statement is allowed per line, but...           True  
20       At most one statement is allowed per line, but...           True  
21       At most one statement is allowed per line, but...           True  
...                                                    ...            ...  
1941503  Define a constant instead of duplicating this ...           True  
1941504  Define a constant instead of duplicating this ...           True  
1941505  Define a constant instead of duplicating this ...           True  
1941506  Define and throw a dedicated exception instead...           True  
1941507  Rename this method name to match the regular e...           True  

[119254 rows x 3 columns]

In [48]:
tags_df2 = check2.drop(columns = ["canBeInferred"]).merge(df_map, how = "inner", left_on = "processedMessage", right_on = "PROCESSED_MESSAGE")
tags_df2 = tags_df2.drop_duplicates()
tags_df2

message  \
0       Define and throw a dedicated exception instead...   
10069   Add a private constructor to hide the implicit...   
10614   At most one statement is allowed per line, but...   
10744   At most one statement is allowed per line, but...   
12140   At most one statement is allowed per line, but...   
...                                                   ...   
120145  Call the static method Short.toString(...) ins...   
120147  Call the static method Double.toString(...) in...   
120149  Call the static method Byte.toString(...) inst...   
120151  Call the static method Float.toString(...) ins...   
120153                        "Argument 1" is a "Thread".   

                                         processedMessage  \
0       Define and throw a dedicated exception instead...   
10069   Add a private constructor to hide the implicit...   
10614   At most one statement is allowed per line, but...   
10744   At most one statement is allowed per line, but...   
12140   At most one statement is allowed per line, but...   
...                                                   ...   
120145  Call the static method Short.toString(...) ins...   
120147  Call the static method Double.toString(...) in...   
120149  Call the static method Byte.toString(...) inst...   
120151  Call the static method Float.toString(...) ins...   
120153                                             is a .   

                           TAGS  \
0       cert,cwe,error-handling   
10069                    design   
10614                     style   
10744                     style   
12140                     style   
...                         ...   
120145                   clumsy   
120147                   clumsy   
120149                   clumsy   
120151                   clumsy   
120153  multi-threading,pitfall   

                                        PROCESSED_MESSAGE  
0       Define and throw a dedicated exception instead...  
10069   Add a private constructor to hide the implicit...  
10614   At most one statement is allowed per line, but...  
10744   At most one statement is allowed per line, but...  
12140   At most one statement is allowed per line, but...  
...                                                   ...  
120145  Call the static method Short.toString(...) ins...  
120147  Call the static method Double.toString(...) in...  
120149  Call the static method Byte.toString(...) inst...  
120151  Call the static method Float.toString(...) ins...  
120153                                             is a .  

[18343 rows x 4 columns]

In [49]:
tags_df2[tags_df2["TAGS"].isnull()]

message  \
95498  The Cyclomatic Complexity of this class is 211...   
95499  The Cyclomatic Complexity of this class is 220...   
95500  The Cyclomatic Complexity of this class is 279...   
95501  The Cyclomatic Complexity of this class is 250...   
95502  The Cyclomatic Complexity of this class is 271...   
...                                                  ...   
95591  The Cyclomatic Complexity of this class is 321...   
95596  The Cyclomatic Complexity of this class is 274...   
95597  The Cyclomatic Complexity of this class is 319...   
95598  The Cyclomatic Complexity of this class is 226...   
95600  The Cyclomatic Complexity of this class is 407...   

                                        processedMessage TAGS  \
95498  The Cyclomatic Complexity of this class is  wh...  NaN   
95499  The Cyclomatic Complexity of this class is  wh...  NaN   
95500  The Cyclomatic Complexity of this class is  wh...  NaN   
95501  The Cyclomatic Complexity of this class is  wh...  NaN   
95502  The Cyclomatic Complexity of this class is  wh...  NaN   
...                                                  ...  ...   
95591  The Cyclomatic Complexity of this class is  wh...  NaN   
95596  The Cyclomatic Complexity of this class is  wh...  NaN   
95597  The Cyclomatic Complexity of this class is  wh...  NaN   
95598  The Cyclomatic Complexity of this class is  wh...  NaN   
95600  The Cyclomatic Complexity of this class is  wh...  NaN   

                                       PROCESSED_MESSAGE  
95498  The Cyclomatic Complexity of this class is  wh...  
95499  The Cyclomatic Complexity of this class is  wh...  
95500  The Cyclomatic Complexity of this class is  wh...  
95501  The Cyclomatic Complexity of this class is  wh...  
95502  The Cyclomatic Complexity of this class is  wh...  
...                                                  ...  
95591  The Cyclomatic Complexity of this class is  wh...  
95596  The Cyclomatic Complexity of this class is  wh...  
95597  The Cyclomatic Complexity of this class is  wh...  
95598  The Cyclomatic Complexity of this class is  wh...  
95600  The Cyclomatic Complexity of this class is  wh...  

[78 rows x 4 columns]

The current dataframe is filled with tags that could directly be inferred from version 2. Still missing are the values that aren't
common with version 2, but still can be inferred from striking substrings.

In [50]:
tags_df2[tags_df2["TAGS"].isnull()]["processedMessage"].unique()

array(['The Cyclomatic Complexity of this class is  which is greater than  authorized.'],
      dtype=object)

In [51]:
check[check["PROCESSED_MESSAGE"].str.contains("TODO")]

TAGS                                            MESSAGE  \
5  cwe  Complete the task associated to this TODO comm...   

                                   PROCESSED_MESSAGE  
5  Complete the task associated to this TODO comm...

In [52]:
check[check["PROCESSED_MESSAGE"].str.contains("Cyclomatic Complexity")]

TAGS                                            MESSAGE  \
8294    brain-overload  The Cyclomatic Complexity of this method "load...   
12854              NaN  The Cyclomatic Complexity of this class is 407...   
445054  brain-overload  The Cyclomatic Complexity of this function "re...   

                                        PROCESSED_MESSAGE  
8294    The Cyclomatic Complexity of this method  is  ...  
12854   The Cyclomatic Complexity of this class is  wh...  
445054  The Cyclomatic Complexity of this function  is...

In [53]:
check[check["PROCESSED_MESSAGE"].str.contains("Take the required action")]

TAGS                                            MESSAGE  \
9850  cwe  Take the required action to fix the issue indi...   

                                      PROCESSED_MESSAGE  
9850  Take the required action to fix the issue indi...

In [54]:
tag_map = {
    "Take the required action": "cwe",
    "Cyclomatic Complexity": "brain-overload",
    "TODO": "cwe"
}

# Iterate through the mapping
for substring, tag in tag_map.items():
    # Create a boolean mask for rows where 'processedMessage' contains the substring
    mask = tags_df2['processedMessage'].str.contains(substring, na=False)

    # Fill NaN values in 'TAGS' for the rows where the mask is True
    tags_df2.loc[mask & tags_df2['TAGS'].isna(), 'TAGS'] = tag

# Display the first few rows with the filled TAGS
tags_df2[tags_df2["TAGS"].isnull()]

Empty DataFrame
Columns: [message, processedMessage, TAGS, PROCESSED_MESSAGE]
Index: []

All tags have been filled.

In [55]:
tags_df2

message  \
0       Define and throw a dedicated exception instead...   
10069   Add a private constructor to hide the implicit...   
10614   At most one statement is allowed per line, but...   
10744   At most one statement is allowed per line, but...   
12140   At most one statement is allowed per line, but...   
...                                                   ...   
120145  Call the static method Short.toString(...) ins...   
120147  Call the static method Double.toString(...) in...   
120149  Call the static method Byte.toString(...) inst...   
120151  Call the static method Float.toString(...) ins...   
120153                        "Argument 1" is a "Thread".   

                                         processedMessage  \
0       Define and throw a dedicated exception instead...   
10069   Add a private constructor to hide the implicit...   
10614   At most one statement is allowed per line, but...   
10744   At most one statement is allowed per line, but...   
12140   At most one statement is allowed per line, but...   
...                                                   ...   
120145  Call the static method Short.toString(...) ins...   
120147  Call the static method Double.toString(...) in...   
120149  Call the static method Byte.toString(...) inst...   
120151  Call the static method Float.toString(...) ins...   
120153                                             is a .   

                           TAGS  \
0       cert,cwe,error-handling   
10069                    design   
10614                     style   
10744                     style   
12140                     style   
...                         ...   
120145                   clumsy   
120147                   clumsy   
120149                   clumsy   
120151                   clumsy   
120153  multi-threading,pitfall   

                                        PROCESSED_MESSAGE  
0       Define and throw a dedicated exception instead...  
10069   Add a private constructor to hide the implicit...  
10614   At most one statement is allowed per line, but...  
10744   At most one statement is allowed per line, but...  
12140   At most one statement is allowed per line, but...  
...                                                   ...  
120145  Call the static method Short.toString(...) ins...  
120147  Call the static method Double.toString(...) in...  
120149  Call the static method Byte.toString(...) inst...  
120151  Call the static method Float.toString(...) ins...  
120153                                             is a .  

[18343 rows x 4 columns]

In [56]:
tags_df2 = tags_df2.drop_duplicates()
tags_df2 = tags_df2.drop(columns = "PROCESSED_MESSAGE")
tags_df2

message  \
0       Define and throw a dedicated exception instead...   
10069   Add a private constructor to hide the implicit...   
10614   At most one statement is allowed per line, but...   
10744   At most one statement is allowed per line, but...   
12140   At most one statement is allowed per line, but...   
...                                                   ...   
120145  Call the static method Short.toString(...) ins...   
120147  Call the static method Double.toString(...) in...   
120149  Call the static method Byte.toString(...) inst...   
120151  Call the static method Float.toString(...) ins...   
120153                        "Argument 1" is a "Thread".   

                                         processedMessage  \
0       Define and throw a dedicated exception instead...   
10069   Add a private constructor to hide the implicit...   
10614   At most one statement is allowed per line, but...   
10744   At most one statement is allowed per line, but...   
12140   At most one statement is allowed per line, but...   
...                                                   ...   
120145  Call the static method Short.toString(...) ins...   
120147  Call the static method Double.toString(...) in...   
120149  Call the static method Byte.toString(...) inst...   
120151  Call the static method Float.toString(...) ins...   
120153                                             is a .   

                           TAGS  
0       cert,cwe,error-handling  
10069                    design  
10614                     style  
10744                     style  
12140                     style  
...                         ...  
120145                   clumsy  
120147                   clumsy  
120149                   clumsy  
120151                   clumsy  
120153  multi-threading,pitfall  

[18343 rows x 3 columns]

In [57]:
tags_df2.to_csv(os.path.join(data_dir, 'tags_message_v1.csv'), index = False)

In [58]:
len(df1["message"].unique())

18417

The mapping for filling tags is almost as long as the unique messages. So for the next step the messages we can fill tags for
have to be matched with through the mapping.

## Filling tags for version 1

In [59]:
df1_tags = df1.merge(tags_df2, how = "inner", on = "message")
df1_tags

projectID        type          creationDate  \
0       commons-daemon  CODE_SMELL  2003-09-04T23:28:19Z   
1       commons-daemon  CODE_SMELL  2003-09-04T23:28:19Z   
2       commons-daemon  CODE_SMELL  2003-09-04T23:28:19Z   
3       commons-daemon  CODE_SMELL  2003-09-27T15:45:02Z   
4       commons-daemon  CODE_SMELL  2010-10-15T19:20:55Z   
...                ...         ...                   ...   
120149     commons-cli  CODE_SMELL  2017-03-25T18:32:55Z   
120150     commons-cli  CODE_SMELL  2017-03-25T18:32:55Z   
120151     commons-cli  CODE_SMELL  2017-07-29T14:36:30Z   
120152     commons-cli  CODE_SMELL  2017-07-29T14:36:30Z   
120153     commons-cli  CODE_SMELL  2017-07-29T14:36:30Z   

                              creationCommitHash closeCommitHash status  \
0       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN   OPEN   
1       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN   OPEN   
2       d3416d3a25b16da3d18b3849522fa96183918e5b             NaN   OPEN   
3       139615aebe97ec81dc22fca5cd2bdd46a1b8cc95             NaN   OPEN   
4       84cf9c0e47fc56ae243f9948ea98297907c13375             NaN   OPEN   
...                                          ...             ...    ...   
120149  269eae18a911f792895d0402f5dd4e7913410523             NaN   OPEN   
120150  269eae18a911f792895d0402f5dd4e7913410523             NaN   OPEN   
120151  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN   OPEN   
120152  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN   OPEN   
120153  3bc9b84d3ae252800eb234a7b41981a9fef8696d             NaN   OPEN   

                                                  message  \
0       Define and throw a dedicated exception instead...   
1       Define and throw a dedicated exception instead...   
2       Define and throw a dedicated exception instead...   
3       Define and throw a dedicated exception instead...   
4       Define and throw a dedicated exception instead...   
...                                                   ...   
120149  Define a constant instead of duplicating this ...   
120150  Define a constant instead of duplicating this ...   
120151  Define a constant instead of duplicating this ...   
120152  Define a constant instead of duplicating this ...   
120153  Define a constant instead of duplicating this ...   

                                         processedMessage  \
0       Define and throw a dedicated exception instead...   
1       Define and throw a dedicated exception instead...   
2       Define and throw a dedicated exception instead...   
3       Define and throw a dedicated exception instead...   
4       Define and throw a dedicated exception instead...   
...                                                   ...   
120149  Define a constant instead of duplicating this ...   
120150  Define a constant instead of duplicating this ...   
120151  Define a constant instead of duplicating this ...   
120152  Define a constant instead of duplicating this ...   
120153  Define a constant instead of duplicating this ...   

                           TAGS  
0       cert,cwe,error-handling  
1       cert,cwe,error-handling  
2       cert,cwe,error-handling  
3       cert,cwe,error-handling  
4       cert,cwe,error-handling  
...                         ...  
120149                   design  
120150                   design  
120151                   design  
120152                   design  
120153                   design  

[120154 rows x 9 columns]

In [60]:
df1_tags[df1_tags["TAGS"].isnull()]

Empty DataFrame
Columns: [projectID, type, creationDate, creationCommitHash, closeCommitHash, status, message, processedMessage, TAGS]
Index: []

In [61]:
len(df1_tags[df1_tags["closeCommitHash"].isnull()])

120154

All issues have been assigned with a tag. Also, since all issues are open, closeCommitHash is always NA and can be removed.

In [62]:
df1_tags = df1_tags.drop(columns = ["processedMessage", "closeCommitHash"])

In [63]:
df1_tags.to_csv(os.path.join(data_dir, 'issues+tags_v1.csv'), index = False)

In [64]:
df_tags2 = df1_tags[["creationCommitHash", "TAGS"]]
df_tags2 = df_tags2.drop_duplicates()

When dropping duplicates based on commit ID and tags, only 27320 rows remain. However, since tags are assigned to messages,
the creationCommitHash values can still appear more than once. Similarly to version 2, the tags need to be flatted per commit.
This means that every commit only has one list of tags that shows which code smells have been newly detected that commit.

In [65]:
df_tags2.groupby(by = "creationCommitHash").count()

TAGS
creationCommitHash                            
0001f90914b418859eb9fa86903e89a793e48e9b     3
0010fcad01ce4ac5051a61bec349bd6ac397c994     1
001e87db6432867d72f59553359d0e4cee5cd31d     1
002e4de0d68ffab9af339283f02659472a1febe7     2
002f40f9d3ec65891b5faf086df404dd3c450600     8
...                                        ...
ffb2c376f08ea23854f4fad4933005b34827d5a7     4
ffcdbb17689ed325715a43dc407ca20348277a5a     1
ffd4863413cdb91db6f21744ada1d0b4a1559626     3
ffd8e16a8a418dce679e07a973d6ff9bb5b304aa     6
ffe13cfcca913a3eff9cec27f0a3ad2b6dcedcc8     4

[9442 rows x 1 columns]

The tags and ID combinations must be prepared similarly to version 2, which means they first get grouped by the ID and then
split into proper lists where each tag is an item instead of each combination of tags being an item.

In [66]:
df_tags2 = df_tags2.groupby('creationCommitHash')['TAGS'].unique().reset_index(name='uniqueTags')
df_tags2

creationCommitHash  \
0     0001f90914b418859eb9fa86903e89a793e48e9b   
1     0010fcad01ce4ac5051a61bec349bd6ac397c994   
2     001e87db6432867d72f59553359d0e4cee5cd31d   
3     002e4de0d68ffab9af339283f02659472a1febe7   
4     002f40f9d3ec65891b5faf086df404dd3c450600   
...                                        ...   
9437  ffb2c376f08ea23854f4fad4933005b34827d5a7   
9438  ffcdbb17689ed325715a43dc407ca20348277a5a   
9439  ffd4863413cdb91db6f21744ada1d0b4a1559626   
9440  ffd8e16a8a418dce679e07a973d6ff9bb5b304aa   
9441  ffe13cfcca913a3eff9cec27f0a3ad2b6dcedcc8   

                                             uniqueTags  
0         [cert,cwe,error-handling, convention, design]  
1                                      [brain-overload]  
2                                          [convention]  
3                     [cert,cwe,error-handling, design]  
4     [cert,cwe,error-handling, brain-overload, anti...  
...                                                 ...  
9437  [brain-overload, bad-practice,cert, misra,unus...  
9438                                           [clumsy]  
9439  [cert,cwe,error-handling, cert,error-handling,...  
9440  [cert,cwe,error-handling, bad-practice,cert,cw...  
9441  [convention, suspicious, antipattern,code-smel...  

[9442 rows x 2 columns]

In [67]:
df_tags2["uniqueTags"] = df_tags2["uniqueTags"].apply(safe_split)
df_tags2

creationCommitHash  \
0     0001f90914b418859eb9fa86903e89a793e48e9b   
1     0010fcad01ce4ac5051a61bec349bd6ac397c994   
2     001e87db6432867d72f59553359d0e4cee5cd31d   
3     002e4de0d68ffab9af339283f02659472a1febe7   
4     002f40f9d3ec65891b5faf086df404dd3c450600   
...                                        ...   
9437  ffb2c376f08ea23854f4fad4933005b34827d5a7   
9438  ffcdbb17689ed325715a43dc407ca20348277a5a   
9439  ffd4863413cdb91db6f21744ada1d0b4a1559626   
9440  ffd8e16a8a418dce679e07a973d6ff9bb5b304aa   
9441  ffe13cfcca913a3eff9cec27f0a3ad2b6dcedcc8   

                                             uniqueTags  
0       [cwe, error-handling, design, convention, cert]  
1                                      [brain-overload]  
2                                          [convention]  
3                   [cwe, error-handling, design, cert]  
4     [cwe, error-handling, brain-overload, antipatt...  
...                                                 ...  
9437  [brain-overload, unused, misra, bad-practice, ...  
9438                                           [clumsy]  
9439  [cwe, error-handling, suspicious, convention, ...  
9440  [cwe, error-handling, brain-overload, clumsy, ...  
9441  [antipattern, clumsy, suspicious, convention, ...  

[9442 rows x 2 columns]

In [68]:
df_tags2.to_csv(os.path.join(data_dir, 'tags_commithash_v1.csv'), index = False)